### PR TITLE
migration events

### LINKED ISSUE #39: Migration Events
Broadcast 'read only' events passing the migration and the respective response/request.

- [x] `RequestIsMigrating`
- [x] `RequestHasMigrated`
- [x] `ResponsetIsMigrating`
- [x] `ResponseHasMigrated`

### PR DESCRIPTION
**Pull request by tomschlick**
Fixes #39 

@sixlive take a look at this and LMK what you think

### CODE REVIEW

**Comment by jbrooksuk** [comment]
👍 from me!

File: src/Migrator.php [hunk 1]
```diff
@@ -110,7 +114,14 @@ public function processRequestMigrations() : Request
             })
             ->flatten()
             ->each(function ($migration) {
-                $this->request = (new $migration())->migrateRequest($this->request);
+                $class = (new $migration());
+                $originalRequest = $this->request;
+
+                event(new RequestIsMigrating($class, $originalRequest));
```
Comment (sixlive) [inline src/Migrator.php]: Maybe use the Facade rather than the helper since that's the pattern we've been using.

File: src/Migrator.php [hunk 2]
```diff
@@ -5,7 +5,11 @@
 use Illuminate\Support\Arr;
 use Illuminate\Http\Request;
 use Illuminate\Support\Collection;
+use Illuminate\Support\Facades\Event;
```
Comment (sixlive) [inline src/Migrator.php:8]: Don't need this line since you used the helper, but I think we should use the facade.

File: src/Migrator.php [hunk 3]
```diff
@@ -110,7 +114,14 @@ public function processRequestMigrations() : Request
             })
             ->flatten()
             ->each(function ($migration) {
-                $this->request = (new $migration())->migrateRequest($this->request);
+                $class = (new $migration());
```
Comment (sixlive) [inline src/Migrator.php]: Do we need the additional encapsulation `$class = new $migration;` ?

File: tests/MigratorTest.php [hunk 4]
```diff
@@ -2,16 +2,44 @@
...
+        Event::fake();
     }
 
     /** @test */
     public function it_will_only_get_versions_behind_current()
     {
         $this->markTestIncomplete();
     }
+
+    protected function assertMigrationEventsFired()
```
Comment (sixlive) [inline tests/MigratorTest.php:26]: Future use for testing I assume?

**Review by sixlive — CHANGES_REQUESTED** [review]
Outside of the few comments, this looks perfect!

File: src/Migrator.php (hunk 2 above)
Comment (jbrooksuk) [reply src/Migrator.php:8]: We should 😉

File: tests/MigratorTest.php (hunk 4 above)
Comment (tomschlick) [reply tests/MigratorTest.php:26]: Yeah. I didn't have the brain power to write more tests last night to generate a fake request / response.

File: src/Migrator.php (hunk 2 above)
Comment (tomschlick) [reply src/Migrator.php:8]: 👍

File: tests/MigratorTest.php (hunk 4 above)
Comment (sixlive) [reply tests/MigratorTest.php:26]: Cool! Totally makes sense!

**Review by sixlive — APPROVED** [review]

**Pull request merged**